README.md
[![Chef cookbook](https://img.shields.io/cookbook/v/postgresql_lwrp.svg)](https://github.com/express42/postgresql_lwrp)
[![Code Climate](https://codeclimate.com/github/express42/postgresql_lwrp/badges/gpa.svg)](https://codeclimate.com/github/express42/postgresql_lwrp)
[![Build Status](https://travis-ci.org/express42/postgresql_lwrp.svg)](https://travis-ci.org/express42/postgresql_lwrp)

Description
===========
This cookbook includes recipes and providers to install and configure postgresql database. This cookbook was tested with Postgresql 9.1, 9.2, 9.3, 9.4, 9.5, 9.6 & 10.

Supported platforms:

* Debian 8
* Debian 9
* Ubuntu 14.04
* Ubuntu 16.04
* Ubuntu 18.04

*Note: TravisCI tests for Ubuntu 18.04 are omitted now because they somehow hang. Local Vagrant & Docker-based tests are succesfull. This will be investigated further.*

Changelog
=========
See [CHANGELOG.md](https://github.com/express42/postgresql_lwrp/blob/master/CHANGELOG.md)

Requirements
============

The minimal recommended version of chef-client is `13.0.113`. It may still work on version `12.5.1` and older, but no tests are made starting from version `1.3.0` of this cookbook as Chef 12 is reaching its EOL in the April, 2018

Dependencies
============

* apt
* cron
* poise-python

Attributes
==========
This cookbook have server and client attribute files.

With client attributes(["postgresql"]["client"]) you can set only postgresql client and library version.

Server attributes are starting from ["postgresql"]["defaults"] and used as default attributes for postgresql provider. You should not override this defaults, you can pass your settings to provider instead.

Resources/Providers
===================

### Resource: default

#### Actions

- :create: creates postgresql cluster

#### Resource parameters

- cluster_name: name attribute. Cluster name (e.g. main). Be aware, systemd (in Ubuntu 16.04 and Debian Jessie) not working with cluster names that containing dashes ('-').
- cluster_version: set cluster version
- cookbook: cookbook for templates. Skip this for default templates.
- cluster_create_options: options for pg_createcluster (only locale related options)
- configuration: Hash with configuration options for postgresql, see examples.
- hba_configuration: Array with hba configuration, see examples.
- ident_configuration: Array with ident configuration, see examples.
- replication: Hash with replication configuration. See replication example.
- replication_initial_copy: Boolean. If `true` pg_basebackup will be exec to make initial replication copy. Default is `false`.
- replication_start_slave: Boolean. If `true` slave cluster will be started after creation. Should be used with replication_initial_copy option. Default `false`.
- allow_restart_cluster: Can be `first`, `always` or `none`. Specifies when cluster must restart instead of reload. `first` – only first time after installation. `always` – always restart, even if changes doesn't require restart. `none` - never, use reload every time. Default is `none`.


Other
=====
### Cloud backup helper:

`postgresql_cloud_backup_helper.sh` helper can be found at `/opt/wal-e/bin/`.

#### Usage:

`postgresql_cloud_backup_helper.sh <cluster_name> <cluster_version> last|count`

- `cluster_name` – postgresql cluster name (ex. *main*)
- `cluser_version` – postgresql cluser version (ex. 9.3)
- `last` – shows last backup time
- `count` – shows total number of backups.

Examples
========
Example master database setup:

```ruby
postgresql 'main' do
  cluster_version '9.3'
  cluster_create_options( locale: 'ru_RU.UTF-8' )
  configuration(
      listen_addresses:           '192.168.0.2',
      max_connections:            300,
      ssl_renegotiation_limit:    0,
      shared_buffers:             '512MB',
      maintenance_work_mem:       '64MB',
      work_mem:                   '8MB',
      log_min_duration_statement: 200
  )
  hba_configuration(
    [
      { type: 'host', database: 'all', user: 'all', address: '192.168.0.0/24', method: 'md5' },
      { type: 'host', database: 'replication', user: 'postgres', address: '192.168.0.3/32', method: 'trust' }
    ]
  )
end
```

Example slave database setup:

```ruby
postgresql 'main' do
   cluster_version '9.3'
  cluster_create_options( locale: 'ru_RU.UTF-8' )
  configuration(
      listen_addresses:           '192.168.0.3',
      max_connections:            300,
      ssl_renegotiation_limit:    0,
      shared_buffers:             '512MB',
      maintenance_work_mem:       '64MB',
      work_mem:                   '8MB',
      log_min_duration_statement: 200
  )
  hba_configuration(
    [
      { type: 'host', database: 'all', user: 'all', address: '192.168.0.0/24', method: 'md5' },
      { type: 'host', database: 'replication', user: 'postgres', address: '192.168.0.2/32', method: 'trust' }
    ]
  )
  replication(
    standby_mode: 'on',
    primary_conninfo: 'host=192.168.0.1',
    trigger_file: '/tmp/pgtrigger'
  )
  replication_initial_copy true
  replication_start_slave true
end
```

Example slave configuration with replication slots (PostgreSQL >= 9.4)

```ruby
replication(
  standby_mode: 'on',
  primary_conninfo: 'host=192.168.0.1',
  trigger_file: '/tmp/pgtrigger'
  primary_slot_name: 'some_slot_on_master'
)
```
Don't forget to create slot on master server before:

```sql
# SELECT pg_create_physical_replication_slot('some_slot_on_master');
```

Example users and databases setup

```ruby
postgresql_user 'user01' do
  in_version '9.3'
  in_cluster 'main'
  unencrypted_password 'user01password'
end

postgresql_database 'database01' do
  in_version '9.3'
  in_cluster 'main'
  owner 'user01'
end
```

Example full daily database backup

```ruby
postgresql_cloud_backup 'main' do
  utility 'wal-g'
  in_version '9.3'
  in_cluster 'main'
  full_backup_time weekday: '*', month: '*', day: '*', hour: '3', minute: '0'
  # Data bag item should contain following keys for S3 protocol:
  # aws_access_key_id, aws_secret_access_key, wale_s3_prefix
  parameters Chef::EncryptedDataBagItem.load('s3', 'secrets').to_hash.select {|i| i != "id"}
  # Or just a hash, if you don't use data bags:
  parameters { aws_access_key_id: 'access_key', aws_secret_access_key: 'secret_key', walg_s3_prefix: 's3_prefix' }
  # In case you need to prepend wal-e with, for example, traffic limiter
  # you can use following method:
  command_prefix 'trickle -s -u 1024'
  # It will be prepended to resulting wal-e execution in cron task
end
```

Example usage of cloud backup helper usage

```bash
$ /opt/wal-e/bin/postgresql_cloud_backup_helper.sh main 9.3 last
1428192159
$ /opt/wal-e/bin/postgresql_cloud_backup_helper.sh main 9.3 count
31
```

Example of how to install extensions from postgresql-contrib
NOTE: schema and version are optional parameters, but others are required

```ruby
postgresql_extension 'cube' do
  in_version '9.4'
  in_cluster 'main'
  db 'test01'
  schema 'public'
end
```
Example of how to install extensions from http://pgxn.org/
NOTE: schema is an optional parameter, but others are required

```ruby
pgxn_extension 'pg_lambda' do
  in_version '9.4'
  in_cluster 'main'
  db 'test01'
  version '1.0.2'
  stage 'stable'
end
```


# License and Maintainer

Maintainer:: LLC Express 42 (<cookbooks@express42.com>)
Source:: https://github.com/express42/postgresql_lwrp
Issues:: https://github.com/express42/postgresql_lwrp/issues

License:: MIT


CHANGELOG.md
## 1.6.3 (Aug 19, 2019)

* (New) wal-g version 0.2.11
* (New) bundle update

## 1.6.2 (Aug 05, 2019)

* (Fix) more typos in attributes

## 1.6.1 (Jul 30, 2019)

* (Fix) typos in attributes.

## 1.6.0 (Jul 03, 2019)

* (Deprecation) `postgresql_cloud_backup` attribute `protocol` was removed. It was used only for a partial env vars validation.

## 1.5.4 (Jun 28, 2019)

* Travis config: tests for Ubuntu 18.04 were removed again. Run local ones.

## 1.5.3 (Jun 27, 2019)

* (New) WAL-G version 0.2.9
* (New) `wal-g wal-push` fix by injecting the `PATH` variable into the envdir
* (New) ruby 2.6.3 for TravisCI
* (New) Test Kitchen suites for Chef 15

## 1.5.2 (Feb 25, 2019)

* (New) tests for PostgreSQL 9.1 & 9.2 were dropped from TravisCI configuration
* (New) WAL-G version 0.2.6 which has new GPG support implementation

## 1.5.1 (Feb 17, 2019)

* (Fix) envdir's files should be marked as sensitive.
* (Fix) Set PGHOST environment variable to `/var/run/postgresql` for wal-g if unset.

## 1.5.0 (Feb 16, 2019)

* (New) PostgreSQL 11 tests
* (New) wal-g support for cloud backup
* (New) Ruby 2.6.0 for TravisCI tests
* (New) Inspec configuration to test running PostgreSQL configuration
* (Fix) pip version 18.0 was pinned as the newer fails to install to sandbox
* (Fix) Some minor ruby style fixes
* (Fix) Other pg extension should be installed during integration tests

## 1.4.2 (May 22, 2018)

* (New) Ubuntu 18.04 tests
* (New) Test Kitchen configuration for AWS cloud was removed
* (Fix) cookbook `poise-python` version was restricted again `>= 1.7.0`
* (Fix) Some minor Test Kitchen configuration improvements

## 1.4.1 (May 15, 2018)

* (Fix) Unpin `poise-python` cookbook version

## 1.4.0 (Apr 10, 2018)

* (New) Chef 14 support & tests
* (New) LWRP-defined resources were rewritten using the new custom resource style
* (Fix) TravisCI: all Chef 13 test were enabled

## 1.3.1 (Apr 03, 2018)
* (Fix) TravisCI build method. Now using `sethvargo/stove` gem instead of `dpl`

## 1.3.0 (Apr 03, 2018)
* (New) Chef 12 test were dropped. (Chef 12 reaches EOL at the and of April 2018)
* (New) Chef 13 full support;
* (New) WAL-E version 1.1.0;
* (New) `poise-python` cookbook is used instead of outdated `python`;
* (Fix) TravisCI configuration was updated to test against more OSes and to use Chef 13;
* (Fix) Berksfile version pins were removed;

## 1.2.4 (Mar 12, 2018)
* (Fix) Rename `params` method in `postgresql_cloud_backup` for compatibility with Chef 13

## 1.2.3 (Jan 31, 2018)
* (Fix) Resource `postgresql` PostgreSQL version validation.
* (Fix) Use resource attributes to set PostgreSQL version for test purposes.

## 1.2.2 (Jan 16, 2018)
* (New) PostgreSQL 10 support.
* (New) Integration tests were migrated to InSpec.
* (New) InSpec resources: postgres_database
* (Fix) InSpec resources: `postgres_cluster`, `postgres_extension` & `postgres_user` were refactored.
* (Fix) Test Kitchen: use one test recipe instead of one-recipe-per-pg-version.
* (Fix) Test Kitchen: use only official images.
* (Fix) Test Kitchen: tests for Chef 11 support were removed, as outdated.
* (Fix) Test Kitchen: Test for Postgresql 9.0 were removed; there is PostgreSQL 9.0 package on modern systems.
* (Fix) [postgresql] fix ruby_block notifications.
* (Fix) [pgtest] user creation should be invoked using `encrypted_password` attribute for better compatibility.

## 1.2.1 (Dec 15, 2016)
* (New) Autoremove checkpoint_segments from configuration if pg > 9.4

## 1.2.0 (Jul 22, 2016)
* (New) Add extension lwrp to install extensions from postgresql-contrib subpackage, which comes installed
* (New) Add pgxn extension lwrp to install extensions from pgxn.org website, using pgxn client
* (New) Add test recipes for installing extensions with newly introduced resources

## 1.1.15 (Sep 24, 2015)
* (Fix) [postgresql] Fix initial slave creation on 9.1

## 1.1.14 (Aug 11, 2015)
* (Fix) [postgresql_database] Fix database existence

## 1.1.13 (Jul 17, 2015)
* (Fix) [common] Fix run under Chef 11

## 1.1.12 (Jun 27, 2015)
* (Fix) [common] Fix compatibility with Chef 12.4.0
* (New) [postgresql_user] Use inline resources in user provider

## 1.1.11 (Apr 7, 2015)
* (New) [common] Return to LR for all resources

## 1.1.10 (Apr 5, 2015)
* (New) [cloud_backup] Add backup retention
* (New) [cloud_backup] Add postgresql_cloud_backup_helper.sh (See README)
* (Fix) [cloud_backup] install libffi-dev package for cffi
* (Fix) [common] Fix reload on Chef 12

## 1.1.9 (Mar 5, 2015)
* (Fix) [metadata] Fix recipes name

## 1.1.8 (Mar 3, 2015)
* (New) [Replication] Add primary_slot_name param support in recovery.conf
* (New) [cloud_backup] Add add prefix to crontab command
* (Fix) [packages] Install dev package only for actual cluster version

## 1.1.7 (Jan 22, 2015)
* (Fix) Remove wal-e pip attribute
* (New) Add serverspec tests for cloud backup

## 1.1.6 (Jan 21, 2015)
* (New) Use virtualenv for wal-e
* (New) Add test recipe for cloud backup

## 1.1.5 (Dec 28, 2014)
* (Fix) Fix cloud_backup cron script name

## 1.1.4 (Dec 22, 2014)
* (Fix) Fix pg version checks

## 1.1.3 (Dec 22, 2014)
* (New) ssl key and cert linkage for pg < 9.2

## 1.1.2 (Dec 22, 2014)
* (Fix) Fix full_backup_time param

## 1.1.1 (Dec 18, 2014)
* (Fix) Fix Test Kitchen boxes
* (Fix) Fix postgresql start after reboot

## 1.1.0 (Dec 10, 2014)
* (New) Add cloud backup lwrp, using wal-e for cloud backup

## 1.0.1 (Oct 31, 2014)
* (Fix) Fix broken allow_restart_cluster option

## 1.0.0 (Aug 25, 2014)

* (New) Flat configuration file
* (New) Initial replicaton can be started automatically
* (New) Option allow_restart_cluster allows do restart instead reload (Only first time or always)
* (New) Resources/providers for database and user creation
* (New) Recipe apt_official_repository with official postgresql repository
* (New) Severspec tests added
* (Removed) Removed databags for users and databases. You should use appropriate providers
* (Fix) pg_ident template fixed

## 0.2.3 (Jun 18, 2013)

### Minor fixes


resources/default.rb
#
# Cookbook Name:: postgresql_lwrp
# Resource:: default
#
# Author:: Kirill Kouznetsov (agon.smith@gmail.com)
#
# Copyright (C) 2014 LLC Express 42
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

include Chef::Postgresql::Helpers

provides :postgresql
resource_name :postgresql

default_action :create

property :cluster_name, String, required: true
property :cluster_version, String, regex: [/\A(1\d|9\.[1-6])\Z\z/]
property :cookbook, String, default: 'postgresql_lwrp'
property :cluster_create_options, Hash, default: {}
property :configuration, Hash, default: {}
property :hba_configuration, Array, default: []
property :ident_configuration, Array, default: []
property :replication, Hash, default: {}
property :replication_initial_copy, [TrueClass, FalseClass], default: false
property :replication_start_slave, [TrueClass, FalseClass], default: false
property :allow_restart_cluster, default: :none, callbacks: {
  'is not allowed! Allowed params for allow_restart_cluster: first, always or none' => proc do |value|
    !value.to_sym.match(/^(first|always|none)$/).nil?
  end,
}

action :create do
  configuration            = Chef::Mixin::DeepMerge.merge(node['postgresql']['defaults']['server']['configuration'].to_hash, new_resource.configuration)
  hba_configuration        = node['postgresql']['defaults']['server']['hba_configuration'] | new_resource.hba_configuration
  ident_configuration      = node['postgresql']['defaults']['server']['ident_configuration'] | new_resource.ident_configuration

  cluster_name             = new_resource.name
  cluster_version          = (!new_resource.cluster_version.empty? && new_resource.cluster_version) || node['postgresql']['defaults']['server']['version']
  service_name             = "postgresql_#{cluster_version}_#{cluster_name}"

  allow_restart_cluster    = new_resource.allow_restart_cluster

  replication              = Chef::Mixin::DeepMerge.merge(node['postgresql']['defaults']['server']['replication'].to_hash, new_resource.replication)
  replication_file         = "/var/lib/postgresql/#{cluster_version}/#{cluster_name}/recovery.conf"
  replication_start_slave  = new_resource.replication_start_slave
  replication_initial_copy = new_resource.replication_initial_copy

  cluster_options          = Mash.new(new_resource.cluster_create_options)
  parsed_cluster_options   = []

  first_time               = pg_installed?("postgresql-#{cluster_version}")

  %w(locale lc-collate lc-ctype lc-messages lc-monetary lc-numeric lc-time).each do |option|
    parsed_cluster_options << "--#{option} #{cluster_options[:locale]}" if cluster_options[option]
  end

  # Locale hack
  if new_resource.cluster_create_options.key?('locale') && !new_resource.cluster_create_options['locale'].empty?
    system_lang = ENV['LANG']
    ENV['LANG'] = new_resource.cluster_create_options['locale']
  end

  # Configuration hacks
  configuration_hacks(configuration, cluster_version)

  # Install postgresql-common package
  package 'postgresql-common'

  file '/etc/postgresql-common/createcluster.conf' do
    content "create_main_cluster = false\n"
    only_if { cluster_version.to_f >= 9.2 }
  end

  # Install postgresql server packages
  %W(postgresql-#{cluster_version} postgresql-contrib-#{cluster_version} postgresql-server-dev-#{cluster_version}).each do |pkg|
    package pkg
  end

  # Install pgxn client to download custom extensions
  package 'pgxnclient'
  package 'build-essential'

  # Return locale
  if new_resource.cluster_create_options.key?('locale') && !new_resource.cluster_create_options['locale'].empty?
    ENV['LANG'] = system_lang
  end

  # Systemd not working with cluster names with dashes
  # see http://comments.gmane.org/gmane.comp.db.postgresql.debian/346
  if systemd_used? && cluster_name.include?('-')
    raise "Sorry, systemd not support cluster names with dashes ('-'), use underscore ('_') instead"
  end

  # Create postgresql cluster directories
  %W(
    /etc/postgresql
    /etc/postgresql/#{cluster_version}
    /etc/postgresql/#{cluster_version}/#{cluster_name}
    /var/lib/postgresql
    /var/lib/postgresql/#{cluster_version}
  ).each do |dir|
    directory dir do
      owner 'postgres'
      group 'postgres'
    end
  end

  directory "/var/lib/postgresql/#{cluster_version}/#{cluster_name}" do
    mode '0700'
    owner 'postgres'
    group 'postgres'
  end

  # Exec pg_cluster create
  execute 'Exec pg_createcluster' do
    command "pg_createcluster #{parsed_cluster_options.join(' ')} #{cluster_version} #{cluster_name}"
    not_if do
      ::File.exist?("/etc/postgresql/#{cluster_version}/#{cluster_name}/postgresql.conf") ||
        !replication.empty?
    end
  end

  # Define postgresql service
  postgresql_service = service service_name do
    action :nothing
    provider Chef::Provider::Service::Simple
    start_command "pg_ctlcluster #{cluster_version} #{cluster_name} start"
    stop_command "pg_ctlcluster #{cluster_version} #{cluster_name} stop"
    reload_command "pg_ctlcluster #{cluster_version} #{cluster_name} reload"
    restart_command "pg_ctlcluster #{cluster_version} #{cluster_name} restart"
    status_command "su -c '/usr/lib/postgresql/#{cluster_version}/bin/pg_ctl \
 -D /var/lib/postgresql/#{cluster_version}/#{cluster_name} status' postgres"
    supports status: true, restart: true, reload: true
  end

  # Ruby block for postgresql smart restart
  ruby_block "restart_service_#{service_name}" do
    action :nothing
    block do
      if !replication.empty? && !replication_start_slave
        run_context.notifies_delayed(Chef::Resource::Notification.new(postgresql_service, :reload, self))
      elsif need_to_restart?(allow_restart_cluster.to_sym, first_time)
        run_context.notifies_delayed(Chef::Resource::Notification.new(postgresql_service, :restart, self))
      else
        run_context.notifies_delayed(Chef::Resource::Notification.new(postgresql_service, :reload, self))
      end
    end
  end

  # Configuration templates

  main_configuration = configuration.dup
  main_configuration.delete('ssl_cert_file') if cluster_version.to_f < 9.2
  main_configuration.delete('ssl_key_file') if cluster_version.to_f < 9.2
  main_configuration.delete('checkpoint_segments') if cluster_version.to_f > 9.4

  # Backu hack:
  # Set `archive_command` option automatically according to the tool of the
  # choise in `postgresql_cloud_backup` resource assosiated with this PG
  # instance if `archive command` config option is set to `:cloud_auto`
  with_run_context :root do
    t = run_context.resource_collection.select do |res|
      res.resource_name == :postgresql_cloud_backup &&
        res.in_cluster == cluster_name &&
        res.in_version == cluster_version
    end

    if !t.empty? &&
       main_configuration['archive_command'] == :cloud_auto
      backup_utility = t.first.utility
      backup_utility_bin = node['postgresql']['cloud_backup'][backup_utility.sub('-', '_')]['bin']
      main_configuration['archive_command'] = "envdir /etc/#{backup_utility}.d/#{cluster_name}-#{cluster_version}/env/ #{backup_utility_bin} wal-push %p"
    end
  end

  template "/etc/postgresql/#{cluster_version}/#{cluster_name}/postgresql.conf" do
    source 'postgresql.conf.erb'
    owner 'postgres'
    group 'postgres'
    mode '0644'
    variables(
      configuration: main_configuration,
      cluster_name: cluster_name,
      cluster_version: cluster_version
    )
    cookbook new_resource.cookbook
    notifies :run, "ruby_block[restart_service_#{service_name}]", :delayed
  end

  template "/etc/postgresql/#{cluster_version}/#{cluster_name}/pg_hba.conf" do
    source 'pg_hba.conf.erb'
    owner 'postgres'
    group 'postgres'
    mode '0644'
    variables configuration: hba_configuration
    cookbook new_resource.cookbook
    notifies :run, "ruby_block[restart_service_#{service_name}]", :delayed
  end

  template "/etc/postgresql/#{cluster_version}/#{cluster_name}/pg_ident.conf" do
    source 'pg_ident.conf.erb'
    owner 'postgres'
    group 'postgres'
    mode '0644'
    variables configuration: ident_configuration
    cookbook new_resource.cookbook
    notifies :run, "ruby_block[restart_service_#{service_name}]", :delayed
  end

  file "/etc/postgresql/#{cluster_version}/#{cluster_name}/start.conf" do
    content "auto\n"
  end

  # Replication
  if !replication.empty?

    if replication_initial_copy
      pg_basebackup_path = "/usr/lib/postgresql/#{cluster_version}/bin/pg_basebackup"
      pg_data_directory = "/var/lib/postgresql/#{cluster_version}/#{cluster_name}"

      BASEBACKUP_PARAMS = {
        'host' => '-h',
        'port' => '-p',
        'user' => '-U',
        'password' => '-W',
      }.freeze

      conninfo_hash = Hash[*replication[:primary_conninfo].scan(/\w+=[^\s]+/).map { |x| x.split('=', 2) }.flatten]

      basebackup_conninfo_hash = conninfo_hash.map do |key, val|
        "#{BASEBACKUP_PARAMS[key.to_s]} #{val}" if BASEBACKUP_PARAMS[key.to_s]
      end.compact

      fetch_wal_logs = if cluster_version.to_i >= 10
                         '-X fetch'
                       else
                         '-x'
                       end

      execute 'Make basebackup' do
        command "#{pg_basebackup_path} -D #{pg_data_directory} -F p #{fetch_wal_logs} -c fast #{basebackup_conninfo_hash.join(' ')}"
        user 'postgres'
        not_if { ::File.exist?("/var/lib/postgresql/#{cluster_version}/#{cluster_name}/base") }
        timeout 604_800
      end
    end

    link "/var/lib/postgresql/#{cluster_version}/#{cluster_name}/server.key" do
      to configuration['ssl_key_file']
      not_if { cluster_version.to_f > 9.1 && ::File.exist?("/var/lib/postgresql/#{cluster_version}/#{cluster_name}/server.key") }
    end

    link "/var/lib/postgresql/#{cluster_version}/#{cluster_name}/server.crt" do
      to configuration['ssl_cert_file']
      not_if { cluster_version.to_f > 9.1 && ::File.exist?("/var/lib/postgresql/#{cluster_version}/#{cluster_name}/server.crt") }
    end

    template "/var/lib/postgresql/#{cluster_version}/#{cluster_name}/recovery.conf" do
      source 'recovery.conf.erb'
      owner 'postgres'
      group 'postgres'
      mode '0644'
      variables replication: replication
      cookbook new_resource.cookbook
      notifies :run, "ruby_block[restart_service_#{service_name}]", :delayed
    end

  else

    file replication_file do
      action :delete
      notifies :run, "ruby_block[restart_service_#{service_name}]", :delayed
    end
  end

  # Start postgresql
  ruby_block "start_service_#{service_name}]" do
    block do
      run_context.notifies_delayed(Chef::Resource::Notification.new(postgresql_service, :start, self))
    end
    not_if { pg_running?(cluster_version, cluster_name) || (!replication.empty? && !replication_start_slave) }
  end
end
